File: 0301-0400/0301-Remove Invalid Parentheses/0301-Remove Invalid Parentheses.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        result = []
        def removeParantheses(s, lastLeft, lastRight, left, right):
            bal = 0
            for i in range(lastLeft, len(s)):
                if s[i] == left:
                    bal += 1
                elif s[i] == right:
                    bal -= 1

                if bal >= 0:
                    continue

                for j in range(lastRight, i + 1):
                    if s[j] == right and (j == lastRight or s[j - 1] != right):
                        removeParantheses(s[:j] + s[j+1:], i, j, left, right)
                
                return

            if left == '(':
                removeParantheses(s[::-1], 0, 0, right, left)
            else:
                result.append(s[::-1])

        removeParantheses(s, 0, 0, '(', ')')
        return result

File: 0301-0400/0301-Remove Invalid Parentheses/0301-Remove Invalid Parentheses.cpp
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        vector<string> result;
        removeParantheses(s, 0, 0, '(', ')', result);
        return result;
    }
    
private:
    void removeParantheses(string s, int lastLeft, int lastRight, char left, char right, vector<string>& result) {
        for (int p = 0, i = lastLeft; i < s.size(); ++i) {
            if (s[i] == left) {
                ++p;
            }
            else if (s[i] == right) {
                --p;
            }
            if (p >= 0) {
                continue;
            }
            for (int j = lastRight; j <= i; ++j) {
                if (s[j] == right && (j == lastRight || s[j - 1] != right)) {
                    removeParantheses(s.substr(0, j) + s.substr(j + 1), i, j, left, right, result);
                }
            }
            
            return;
        }
        
        
        if (left == '(') {
            string revs(s);
            reverse(revs.begin(), revs.end());
            removeParantheses(revs, 0, 0, right, left, result);
        }
        else {
            result.push_back(revs);
        }
    }
};
